src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenshotCapture.cs

```diff
@@ -320,51 +320,58 @@ private IEnumerator UpdateGPUData(int frame, Action<(byte[], int, int)?> onCompl
                 }
 
                 var theGraphicsFormat = _screenShotTexture.graphicsFormat;
-
-                try
+                
+                // If we are running in -nographics mode, the async task fails, causing an exception inside
+                // the AsyncGPUReadback.Request that is difficult to catch. This ensures that the screenshot
+                // is only taken when we have graphics
+                if (SystemInfo.graphicsDeviceType != GraphicsDeviceType.Null)
```

Comment:
Review this with whitespace changes turned off, all that changed here is wrapping it in this if statement

Reply:
This code is in the wrong place.  In the current location, the GPUReadbackRequests structure will fill up with key-null associations and not get cleaned up until the end of the recording when WaitForCompletion is called.  It would be better to position this code so that the readback requests are never even tracked, more up towards `WaitForEndOfFrame` in this method.

Reply:
Sorry for taking so long to respond to this - TLDR; I was worried that putting it earlier would have implications in other parts of the code that expect the readback requests to have data. But I _think_ where I put it now is fine, but it would be good to have confirmation.

The other holdup is now everything works as expected, I can run a sequence in headless mode in BossRoom for instance... but all of the zip folders are empty, including the data folders. This is something I need to look into now.

Reply:
(Note: I plan on debugging this myself, I just haven't had a chance to yet. I am going to look into the CompletedActions list and see if those are being completed correctly - if not, that might result in a failure that effects the process in ScreenRecorder.cs which writes the tick data dependent on those callbacks / results.)

Reply:
if that is what you run into, i'd still vote for doing this check early and just calling that completed action right away